rstest_macros/src/render/mod.rs

```diff
@@ -200,19 +203,32 @@ pub(crate) fn matrix(mut test: ItemFn, mut info: RsTestInfo) -> TokenStream {
                     &info,
                     &c.info,
                 )
-                .wrap_by_mod(&c.ident)
+                .map(|recur| recur.wrap_by_mod(&c.ident))
             })
-            .collect()
+            .collect::<Result<_, _>>()?
     };
 
-    test_group(test, rendered_cases)
+    Ok(test_group(test, rendered_cases))
 }
 
-fn resolve_default_test_attr(is_async: bool) -> TokenStream {
-    if is_async {
-        quote! { #[async_std::test] }
+fn resolve_test_attr(
+    is_async: bool,
+    explicit_test_attr: Option<TokenStream>,
+    attributes: &[Attribute],
+    error_span: Span,
+) -> syn::Result<Option<TokenStream>> {
+    if !is_async {
+        Ok(Some(quote! { #[test] }))
+    } else if let Some(explicit_attr) = explicit_test_attr {
+        Ok(Some(explicit_attr))
+    } else if attributes
+        .iter()
+        .any(|attr| attr_ends_with(attr, &parse_quote! {test}))
+    {
+        // test attr is already in the attributes; we don't need to re-inject it
+        Ok(None)
```

Comment:
There is something wrong in this logic: It doesn't work if you want to use a custom test attribute for a not async test.

I'm wondered that there isn't any test that catch it :cry:

Reply:
Good point on the logic.

I haven't started working on testing this yet; I'll add a relevant case.